README.md
# security-final-project

history.txt
Arvind 0
Noob 0
Trump 0
Bobby 0


client.py
import socket
import sys, os
from Crypto.Cipher import PKCS1_OAEP
from Crypto.PublicKey import RSA
from random import randint
from Crypto.Signature import PKCS1_v1_5
from Crypto.Hash import SHA


class Client:

	def __init__(self, host_ip, port_no):
		'''Initialization parameters'''

		self.host_ip = host_ip;
		self.port_no = port_no;

		self.name = ""
		self.random_number = randint(1, 999999)
		self.state = 0

		server_key = open("server_key.pem","r").read()
		server_key_opened = RSA.importKey(server_key)
		server_public_key = server_key_opened.publickey()	# We use the public key of server for encryption
		self.server_cipher = PKCS1_OAEP.new(server_public_key)
		self.client_cipher = self.generate_new_key()



	def establish_connection(self):

		self.soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM);
		
		try:
			self.soc.connect((self.host_ip, self.port_no));

		except:
			print ("[Error] Connection could not be established")
			sys.exit();


	def vote(self):

		details = input("Enter your name and voter_id number seperated by space >> ")
		first_cipher_text = self.process_details(details)
		self.soc.sendall(first_cipher_text)

		if self.soc.recv(1024).decode("utf8") == "0":
			print ("Invalid name or number")
			self.soc.close()
			return True


		print ("Welcome, " + self.name + "\n")
		print ("\tMain Menu\n\nPlease enter a number(1-4)\n1. Vote\n2. My vote history\n3. Election result\n4. Quit\n")
		msg = input(">>> ");

		while msg != ("quit" or "q"):

			enc_msg = self.server_cipher.encrypt(msg.encode("utf8"))
			self.soc.sendall(enc_msg)
			self.state += int(msg)
			from_server = self.soc.recv(1024)
			self.process_input(from_server)
			msg = input("Enter: ");

		# Terminate the connection after one transfer
		self.soc.send(b'quit');


	def process_details(self, details):

		enc_details = self.server_cipher.encrypt(details.encode("utf8"))
		self.name = details.split()[0]
		os.system("cp %d.pem %s.pem" %(self.random_number, self.name))
		os.system("rm %d.pem" %self.random_number)
		key = RSA.importKey(open(self.name+".pem").read())
		_hash = SHA.new(self.name.encode("utf8"))
		signer = PKCS1_v1_5.new(key)
		signature = signer.sign(_hash)
		print ("The length of the details is " + str(len(enc_details)) + " and length of signature is " + str(len(signature)))
		final_message = enc_details + signature
		return final_message


	def generate_new_key(self):

		new_client_key = RSA.generate(2048)
		new_file = open(str(self.random_number)+".pem", "wb")
		new_file.write(new_client_key.exportKey("PEM"))
		new_file.close()
		return PKCS1_OAEP.new(new_client_key)


	def process_input(self, msg):

		pass




def main():

	host = "127.0.0.1"
	port = 54321

	print ("Host:%s Port:%d" %(host, port))
	cli = Client(host, port)

	cli.establish_connection()
	cli.vote()



if __name__ == "__main__":
	main()




'''
vname = input("Enter name: ")
    #soc.sendall(vname.encode("utf8"))

    #Accept reg num and send to server:
    vregnum = input("Enter registration number: ")
    #soc.sendall(vregnum.encode("utf8"))
    vinfo = vname + vregnum
    vinfo_encoded = str.encode(vinfo)
    enc_cipher = PKCS1_OAEP.new(public_key)
    enc_info = enc_cipher.encrypt(vinfo_encoded)

    print("Encrypted data: {}".format(enc_info))

    dec_cipher = PKCS1_OAEP.new(private_key)
    dec_info_encoded = dec_cipher.decrypt(enc_info)
    dec_info = dec_info_encoded.decode()
    print("Decrypted data: {}".format(dec_info))

'''

server.py
import socket
import sys
import traceback
from threading import Thread
from Crypto.Cipher import PKCS1_OAEP
from Crypto.PublicKey import RSA
from Crypto.Signature import PKCS1_v1_5
from Crypto.Hash import SHA


class Server:

	def __init__(self, host_no, port_no, max_people=5):
		'''Initialize the host ip and port number'''

		self.host_no = host_no
		self.port_no = port_no
		
		# Parameters
		self.people_voted = 0
		self.max_people = max_people
		self.people = self.generate_map()
		self.state = 0

		# Socket
		self.soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		
		# Server cipher
		key = open("server_key.pem","rb").read()		
		self.key = RSA.importKey(key)
		self.server_cipher = PKCS1_OAEP.new(self.key)


	def establish_server(self):

		self.soc.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
		print ("Server established")

		try:
			self.soc.bind((self.host_no, self.port_no))

		except:
			print ("[Error] Bind failed")
			sys.exit()

		self.soc.listen(5)
		print ("Socket waiting to accept connections")

		while True:

			connection, address = self.soc.accept()
			ip, port = str(address[0]), str(address[1])

			try:
				Thread(target=self.client_thread, args=(connection, ip, port)).start()

			except:
				print ("Not able to spawn a new thread")
				traceback.print_exc()

		self.soc.close()


	def client_thread(self, connection, ip, port, max_buffer_size = 4096):

		print ("We are inside the client")

		approved = self.authenticate(connection.recv(4096))

		if not approved:
			connection.sendall("0".encode("utf8"))
		else:
			connection.sendall("1".encode("utf8")) 

		is_active = True

		while is_active:

			try:
				client_input = self.process(connection.recv(4096))

			except:
				connection.close()
				break


			if client_input == "4":
				connection.close()
				print("Connection with " + ip + " : " + port + " is now closed")
				is_active = True

			else:
				print ("Processed result : {}".format(client_input))
				connection.sendall("-".encode("utf8"))

		return True


	def authenticate(self, __input):

		client_input = __input
		details, signature = client_input[:256], client_input[256:]
		decoded_input = self.server_cipher.decrypt(details)
		decoded_input = decoded_input.decode("utf8")
		name, v_number = decoded_input.split()
		try:
			assert(self.people[name] == v_number)
		except:
			return False

		key = RSA.importKey((open(name+".pem", "r")).read())
		key = key.publickey()
		_hash = SHA.new(name.encode("utf8"))
		verifier = PKCS1_v1_5.new(key)
		return verifier.verify(_hash, signature)



	def process(self, client_input):

		pass


	def generate_map(self):

		_file = open("voterinfo.txt", "r")
		_hist = open("history.txt", "r")
		v_info = dict()
		voted = dict()

		cur_str = _file.readlines().split()
		cur_hst = _file.readlines().split()

		for n, i in enumerate(cur_str[0]):
			v_info[i] = cur_str[1][n]
			voted[i] = cur_hst[1][n]


		# while (cur_str is not ""):
		# 	name, num = cur_str.split()
		# 	_, hist = cur_hst.split()
		# 	print ("We are now reading ", name)
		# 	v_info[name] = num
		# 	v_hist[name] = hist
		# 	cur_str = _file.readline()
		# 	cur_hst = _hist.readline()

		print("Total length : ", len(v_hist.keys()))

		for i in v_info.keys():
			print (i, v_info[i], v_hist[i])

		return v_info





def main():

	host_no = "127.0.0.1"
	port_no = 54321

	serv = Server(host_no, port_no)
	serv.establish_server()



if __name__ == "__main__":
    main()



generate_keys.py
# Generate keys for the server and client


from Crypto.PublicKey import RSA

# Generate key pair for the server.
serv_key = RSA.generate(2048)
f_serv = open("server_key.pem", "w")
f_serv.write(serv_key.exportKey("PEM"))
f_serv.close()

# Generate key pair for client (Depricated)
# cli_key = RSA.generate(2048)
# f_cli = open("client_key.pem", "w")
# f_cli.write(cli_key.exportKey("PEM"))
# f_cli.close()


vinfo = open("voterinfo.txt", "w")
vhist = open("history.txt", "w")
vinfo.write("Arvind 1994\n")
vhist.write("Arvind 0\n")
vinfo.write("Noob 1995\n")
vhist.write("Noob 0\n")
vinfo.write("Trump 1996\n")
vhist.write("Trump 0\n")
vinfo.write("Bobby 1997\n")
vhist.write("Bobby 0\n")
vinfo.close()
vhist.close()


# Various states
# 0. Main Menu
# 1. Voting Menu
# 2. Voting History Menu
# 3.Election Result


client_with_each_client_sending_own_keys.py
import socket
import sys
from Crypto.Cipher import PKCS1_OAEP
from Crypto.PublicKey import RSA
from Crypto.Hash import SHA256



class Client:

	def __init__(self, host_ip, port_no):
		'''Initialization parameters'''

		self.host_ip = host_ip;
		self.port_no = port_no;
		
		key = open("server_key.pem","r").read()
		self.key = RSA.importKey(key)
		self.public_key = self.key.publickey()	# We use the public key of server for encryption
		key = open("client_key.pem", "r").read()
		self.key = RSA.generate(2048)	# Now we use this key for decryption
		self.own_key = self.key.publickey()
		self.own_key = self.own_key.exportKey()
		self.name = ""
		self.cipher = PKCS1_OAEP.new(self.public_key)
		# print ("The type of cipher is " + str(type(self.public_key)))


	def establish_connection(self):

		self.soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM);
		
		try:
			self.soc.connect((self.host_ip, self.port_no));

		except:
			print ("[Error] Connection could not be established")
			sys.exit();



	def vote(self):

		# Get input from user and send the encrypted version to the server

		self.soc.sendall(self.own_key)
		initial = input("Enter your name and voter number seperated by space>> ")
		to_send = self.prepare_handshake(initial)
		self.soc.sendall(to_send)

		# Verify if the server has approved the request.
		response = self.soc.recv(4096)
		response = self.cipher.decrypt(response)
		response = response.decode("utf8")

		if (response == "0"):
			print ("Invalid name or registration number")
			return 1


		print ("Welcome, " + self.name + "\n")
		print ("\tMain Menu\n\nPlease enter a number(1-4)\n1. Vote\n2. My vote history\n3. Election result\n4. Quit\n")
		msg = input("Enter shit: ");

		while msg != ("quit" or "q"):

			# Encrypt the message using public key of the server.
			enc_msg = self.cipher.encrypt(msg.encode("utf8"))

			self.soc.sendall(enc_msg)
			
			# response = self.soc.recv(1024)
			# cipher = PKCS1_OAEP.new(self.key)
			# response = cipher.decrypt(client_input)
			# response = client_input.decode("utf8").rstrip()

			msg = input(">>>: ");

		# Terminate the connection after one transfer
		self.soc.send(b'quit');



	def prepare_handshake(self, initial_string):
		'''Prepare the initial handshake that concatenates name, number
		   and sends it encrypted with digital signature to the server.'''


		# Concatenate name and v_number with public key of server.
		self.name, v_number = initial_string.split()
		hs_string = self.name + "|" + v_number
		enc_hs_string = self.cipher.encrypt(hs_string.encode("utf8"))

		# Hash name and digitally sign it using private key.
		hasher = SHA256.new()
		hasher.update(self.name.encode("utf8"))
		hashed_name = hasher.hexdigest()
		p_cipher = PKCS1_OAEP.new(self.key)
		ds_hash = p_cipher.encrypt(hashed_name.encode("utf8"))

		# Finally, concatenate name, v_number and digital signature
		to_send = str(enc_hs_string) + "||" + str(ds_hash)
		to_send = to_send.encode("utf8")

		for_now = to_send.decode("utf8")
		_id, _ds = for_now.split("||")

		print ("Id is " + _id + " and DS is " + _ds)
		dec_id = PKCS1_OAEP.new(self.key.publickey).decrypt(_id)

		return to_send



def main():

	host = "127.0.0.1"
	port = 54321

	print ("Host:%s Port:%d" %(host, port))
	cli = Client(host, port)

	cli.establish_connection()

	exit_code = 1

	while (exit_code != 0):
		cli.vote()



if __name__ == "__main__":
	main()




'''
vname = input("Enter name: ")
    #soc.sendall(vname.encode("utf8"))

    #Accept reg num and send to server:
    vregnum = input("Enter registration number: ")
    #soc.sendall(vregnum.encode("utf8"))
    vinfo = vname + vregnum
    vinfo_encoded = str.encode(vinfo)
    enc_cipher = PKCS1_OAEP.new(public_key)
    enc_info = enc_cipher.encrypt(vinfo_encoded)

    print("Encrypted data: {}".format(enc_info))

    dec_cipher = PKCS1_OAEP.new(private_key)
    dec_info_encoded = dec_cipher.decrypt(enc_info)
    dec_info = dec_info_encoded.decode()
    print("Decrypted data: {}".format(dec_info))

'''

server_with_each_client_sending_own_keys.py
import socket
import sys
import traceback
from threading import Thread
from Crypto.Cipher import PKCS1_OAEP
from Crypto.PublicKey import RSA
import os


class Server:

	def __init__(self, host_no, port_no):
		'''Initialize the host ip and port number'''

		self.host_no = host_no
		self.port_no = port_no
		self.soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		os.system("rm *pem")
		os.system("openssl genpkey -algorithm RSA -out server_private_key.pem -pkeyopt rsa_keygen_bits:2048")
		os.system("openssl rsa -pubout -in server_private_key.pem -out server_public_key.pem")
		server_key = open("server_private_key.pem","r").read()
		self.server_private_key = RSA.importKey(server_key)	# Now we use this key for decryption
		self.private_cipher = PKCS1_OAEP.new(self.server_private_key)



	def establish_server(self):

		self.soc.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
		print ("Server established")

		try:
			self.soc.bind((self.host_no, self.port_no))

		except:
			print ("[Error] Bind failed")
			sys.exit()

		self.soc.listen(5)
		print ("Socket waiting to accept connections")

		while True:

			connection, address = self.soc.accept()
			ip, port = str(address[0]), str(address[1])

			try:
				Thread(target=self.client_thread, args=(connection, ip, port)).start()

			except:
				print ("Not able to spawn a new thread")
				traceback.print_exc()

		self.soc.close()



	def client_thread(self, connection, ip, port, max_buffer_size = 4096):


		session_key = self.identify_client(connection.recv(max_buffer_size))
		connection.sendall("1".encode())

		client_input = connection.recv(max_buffer_size)
		self.finish_handshake(client_input, session_key)

		if (ds_verified):
			message = "1"
			connection.sendall(message)

		is_active = True

		while is_active:

			client_input = self.receive_input(connection, max_buffer_size)

			# split_string = 

			if client_input == "4":
				connection.close()
				print("Connection with " + ip + " : " + port + " is now closed")
				is_active = True

			else:
				print ("Processed result : {}".format(client_input))
				connection.sendall("-".encode("utf8"))



	def receive_input(self, connection, max_buffer_size):

		client_input = connection.recv(max_buffer_size)
		# cipher = PKCS1_OAEP.new(self.key)
		decoded_input = self.private_cipher.decrypt(client_input)
		decoded_input = decoded_input.decode("utf8").rstrip()  # decode and strip end of line

		return str(decoded_input)


	def process_input(self, input_str):

	    pass


	def identify_client(self, client_input):
		'''Create a client key object'''

		client_input = client_input.decode()
		return RSA.importKey(client_input)

	
	def finish_handshake(self, client_input, key):

		client_cipher = PKCS1_OAEP.new(key)
		decoded = client_input.decode().rstrip()
		_id, _ds = decoded.split(",,")
		_id = self.private_cipher.decrypt(_id)
		_id = _id.decode()

		
		_ds = self.public_key.decrypt(_ds)
		_name, _vnumber = _id.split(",")
		print ("Name is : " + _name + " and vnumber is " + _vnumber)








def main():

	host_no = "127.0.0.1"
	port_no = 54321

	serv = Server(host_no, port_no)
	serv.establish_server()



if __name__ == "__main__":
    main()



new_client.py
import socket
import sys
from Crypto.Cipher import PKCS1_OAEP
from Crypto.PublicKey import RSA
from Crypto.Hash import SHA256
import time



class Client:

	def __init__(self, host_ip, port_no):
		'''Initialization parameters'''

		self.host_ip = host_ip;
		self.port_no = port_no;
		
		server_key_file = open("server_public_key.pem","r").read()
		server_key = RSA.importKey(server_key_file)
		self.server_public_key = PKCS1_OAEP.new(server_key)	

		client_key = RSA.generate(2048)				# Generate a new key for client
		self.client_public_key = client_key.publickey()
		self.client_public_key_string = self.client_public_key.exportKey()
		self.name = ""
		self.signature = PKCS1_OAEP.new(client_key)


	def establish_connection(self):

		self.soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		
		try:
			self.soc.connect((self.host_ip, self.port_no))

		except:
			print ("[Error] Connection could not be established")
			sys.exit();



	def vote(self):

		# Get input from user and send the encrypted version to the server

		self.soc.sendall(self.client_public_key_string)
		ack = self.soc.recv(4096)
		if (ack.decode() == "0"):
			return 1

		initial = input("Enter your name and voter number seperated by space>> ")
		to_send = self.prepare_handshake(initial)
		self.soc.sendall(to_send)
		time.sleep(50)

		# Verify if the server has approved the request.
		response = self.soc.recv(4096)
		response = self.cipher.decrypt(response)
		response = response.decode("utf8")

		if (response == "0"):
			print ("Invalid name or registration number")
			return 1


		print ("Welcome, " + self.name + "\n")
		print ("\tMain Menu\n\nPlease enter a number(1-4)\n1. Vote\n2. My vote history\n3. Election result\n4. Quit\n")
		msg = input("Enter shit: ");

		while msg != ("quit" or "q"):

			# Encrypt the message using public key of the server.
			enc_msg = self.cipher.encrypt(msg.encode("utf8"))

			self.soc.sendall(enc_msg)
			
			# response = self.soc.recv(1024)
			# cipher = PKCS1_OAEP.new(self.key)
			# response = cipher.decrypt(client_input)
			# response = client_input.decode("utf8").rstrip()

			msg = input(">>>: ");

		# Terminate the connection after one transfer
		self.soc.send(b'quit');



	def prepare_handshake(self, initial_string):
		'''Prepare the initial handshake that concatenates name, number
		   and sends it encrypted with digital signature to the server.'''


		# Concatenate name and v_number with public key of server.
		name, v_number = initial_string.split()
		hs_string = name + "," + v_number
		hs_string = str.encode(hs_string)
		enc_hs_string = self.server_public_key.encrypt(hs_string)

		# Hash name and digitally sign it using private key.
		hasher = SHA256.new()
		hasher.update(str.encode(name))
		hashed_name = hasher.hexdigest()
		ds_hash = self.signature.encrypt(str.encode(hashed_name))

		# Finally, concatenate name, v_number and digital signature
		to_send = str(enc_hs_string) + ",," + str(ds_hash)
		to_send = str.encode(to_send)

		# for_now = to_send.decode("utf8")
		# _id, _ds = for_now.split("||")

		# print ("Id is " + _id + " and DS is " + _ds)
		# dec_id = PKCS1_OAEP.new(self.key.publickey).decrypt(_id)

		return to_send



def main():

	host = "127.0.0.1"
	port = 54321

	print ("Host:%s Port:%d" %(host, port))
	cli = Client(host, port)

	cli.establish_connection()

	exit_code = 1

	while (exit_code != 0):
		cli.vote()



if __name__ == "__main__":
	main()




'''
vname = input("Enter name: ")
    #soc.sendall(vname.encode("utf8"))

    #Accept reg num and send to server:
    vregnum = input("Enter registration number: ")
    #soc.sendall(vregnum.encode("utf8"))
    vinfo = vname + vregnum
    vinfo_encoded = str.encode(vinfo)
    enc_cipher = PKCS1_OAEP.new(public_key)
    enc_info = enc_cipher.encrypt(vinfo_encoded)

    print("Encrypted data: {}".format(enc_info))

    dec_cipher = PKCS1_OAEP.new(private_key)
    dec_info_encoded = dec_cipher.decrypt(enc_info)
    dec_info = dec_info_encoded.decode()
    print("Decrypted data: {}".format(dec_info))

'''